src/pool.jl

```diff
@@ -75,25 +75,28 @@ function mergelevels(ordered, levels...)
         return res, ordered
     end
 
-    for l in levels
-        levelsmap = indexin(l, res)
-
-        i = length(res)+1
-        for j = length(l):-1:1
-            @static if VERSION >= v"0.7.0-DEV.3627"
-                if levelsmap[j] === nothing
-                    insert!(res, i, l[j])
-                else
-                    i = levelsmap[j]
-                end
+    union!(res, levels...)
```

Comment:
Since we are breaking anyway I would just opt for using `union!`. How much cost it the level order fixing incur? If it is a majority of time I would skip doing it and document that we use `union!` when merging levels.

Reply:
(I am asking because it looks like the complexity of the level fixing algorithm below is quadratic)

Reply:
Removing the `for levs in reverse(levels)` loop below gets rid of about 2/3 of the time. That proportion seems to stay roughly constant when moving from 10^4 to 10^7.

But I'm not sure `union!` is really a satisfying solution. Using it would mean that even if the second set of levels is a superset of the first one, we won't keep the order of the superset. Of course that would have the advantage that existing refs wouldn't have to be recoded since new levels would always be added at the end. But the downside is that if you assign categorical values from arrays with different levels (or call `copyto!`), the final levels would depend on the order in which you make the calls.

An intermediate solution would be to check whether one set of levels starts with or ends with a subsequence of another, or is an order superset of another (that should be cheap). If that's the case we can combine them in the appropriate order. If they are not, we can call `union!`.

Reply:
I was thinking about it and concluded that:
1. If the cost is 2/3 of time (and does not explode) then it is acceptable.
2. We should clearly document what we do (and actually highlight that this is an advantage of our approach)
3. We should clearly document that CategoricalArrays.jl is not intended to handle huge numbers of levels (as a lot of book-keeping has to be done)
4. Regarding the implementation. As you suggest what could be checked for the case if we merge two levels (a common case) and one level is a subsequence of another then we probably can be faster (just like you now check if all levels are equal). Such check could be done before checking for the equality.

Reply:
@ablaom Do you have any preference regarding how levels from different pools are merged? Is this something that you encountered in real use cases?